pennylane_forest/converter.py

```diff
@@ -0,0 +1,626 @@
+                gate_matrix = _controlled_matrix(gate_matrix)
+
+        return gate_matrix
+
+    def _check_variable_map(self, variable_map):
+        """Check that all variables of the program are defined.
+
+        Only variables used in measurements need not be defined.
+        
+        Args:
+            variable_map (Dict[str, object]): Map that assigns values to variables
+        
+        Raises:
+            qml.DeviceError: When not all variables are defined in the variable map
+        """
+        for declaration in self._declarations:
+            if not declaration.name in variable_map:
+                # If the variable is used in measurement we don't complain
+                if not declaration.name in self._measurement_variable_names:
+                    raise qml.DeviceError(
+                        (
+                            "The PyQuil program defines a variable {} that is not present in the given variable map. "
+                            + "Instruction: {}"
+                        ).format(declaration.name, declaration)
+                    )
+
+    @property
+    def defined_gates(self):
+        """The custom gates defined in the pyquil Program.
+        
+        Returns:
+            List[Union[pyquil.quil.DefGate, pyquil.quil.DefPermutationGate]]: The gates defined in the pyquil Program
+        """
+        return self.program.defined_gates
+
+    @property
+    def defined_gate_names(self):
+        """The names of the custom gates defined in the pyquil Program.
+
+        Returns:
+            List[str]: The names of the gates defined in the pyquil Program
+        """
+        return self._defined_gate_names
+
+    @property
+    def declarations(self):
+        """The declarations in the pyquil Program.
+
+        Returns:
+            List[pyquil.quil.Declaration]: The declarations in the pyquil Program
+        """
+        return self._declarations
+
+    @property
+    def defined_variable_names(self):
+        """The names of the variables defined in the pyquil Program
+        
+        Returns:
+            List[str]: The names of the variables defined in the pyquil Program
+        """
+        return [declaration.name for declaration in self._declarations]
+
+    def _load_template(self):
+        """Load the template corresponding to the pyquil Program.
+        
+        Raises:
+            qml.DeviceError: When the import of a forked gate is attempted
+        """
+        self._parametrized_gates = []
+
+        for i, instruction in enumerate(self.program.instructions):
+            # Skip all statements that are not gates (RESET, MEASURE, PRAGMA, ...)
+            if not _is_gate(instruction):
+                if not _is_declaration(instruction):
+                    warnings.warn(
+                        "Instruction Nr. {} is not supported by PennyLane and was ignored: {}".format(
+                            i + 1, instruction
+                        )
+                    )
```

Comment:
There is a lot here that is skipped that could change the program output, this is due to the fact that quil does (in theory but not on the hardware) support classical control flow (like MOVE and JUMP and so on), all of which is not differentiable. But we can import everything that is differentiable.
Also the pragmas can be neglected, because they have no relevance for pennylane.

Reply:
Can one still perform noise simulation, if pragmas are neglected? In pyQuil, we can do this by using `ADD-KRAUS` and `READOUT-POVM` pragmas, which essentially translate to converting a deterministic outcome to a probabilistic one for applying  a gate (in the case of kraus) or measurement (in the case of povm). But also, I'm a bit confused how `test_convert_program_with_pragmas` passes while seeming to respect pragmas, if pragmas are indeed neglected -- are the pragmas somehow redundant in that test?

Reply:
No, as it is not supported by pennylane we also don't support importing it. We simply ignore all the pragmas (basically we ignore everything that is neither Gate nor Declaration), and this is also what is tested in that test.